x-pack/legacy/plugins/ml/public/data_frame/pages/job_management/components/job_list/preview_pane.tsx

```diff
@@ -0,0 +1,166 @@
+    // make sure groupBy fields are always most left columns
+    if (groupByArr.some(aggName => aggName === a) && groupByArr.some(aggName => aggName === b)) {
+      return a.localeCompare(b);
+    }
+    if (groupByArr.some(aggName => aggName === a)) {
+      return -1;
+    }
+    if (groupByArr.some(aggName => aggName === b)) {
+      return 1;
+    }
+    return a.localeCompare(b);
+  };
+}
+
+function getDataFromTransform(
+  transformData: TransformData
+): { previewRequest: DataFramePreviewRequest; groupByArr: string[] | [] } {
+  const { transforms } = transformData;
+  const index = transforms[0].source.index[0];
+  const pivot = transforms[0].pivot;
+  const groupByArr = [];
+
+  const previewRequest: DataFramePreviewRequest = {
+    source: {
+      index,
+    },
+    pivot,
+  };
+  // hasOwnProperty check to ensure only properties on object itself, and not its prototypes
+  if (pivot.group_by !== undefined) {
+    for (const key in pivot.group_by) {
+      if (pivot.group_by.hasOwnProperty(key)) {
+        groupByArr.push(key);
+      }
+    }
+  }
+
+  return { groupByArr, previewRequest };
+}
+
+export const PreviewPane: React.SFC<Props> = ({ transformId, lastUpdate }) => {
+  const [dataFramePreviewData, setDataFramePreviewData] = useState([]);
+  const [groupByArray, setGroupByArray] = useState<string[] | []>([]);
+  const [columns, setColumns] = useState<Column[] | []>([]);
+  const [sort, setSort] = useState({});
+  const [isLoading, setIsLoading] = useState(true);
+  const [errorMessage, setErrorMessage] = useState('');
+
+  async function getPreview() {
+    try {
+      const transformData: any = await ml.dataFrame.getDataFrameTransforms(transformId);
+      const { previewRequest, groupByArr } = getDataFromTransform(transformData);
+      const resp: any = await ml.dataFrame.getDataFrameTransformsPreview(previewRequest);
+      setGroupByArray(groupByArr);
+      setDataFramePreviewData(resp.preview);
+    } catch (error) {
+      setIsLoading(false);
+      setErrorMessage(
+        i18n.translate('xpack.ml.dfJobsList.jobDetails.previewPane.errorMessage', {
+          defaultMessage: 'Preview could not be loaded',
+        })
+      );
+    }
+  }
+
+  useEffect(
+    () => {
+      getPreview();
+    },
+    [lastUpdate]
+  );
+  useEffect(
+    () => {
+      if (dataFramePreviewData.length > 0) {
+        const columnKeys = getFlattenedFields(dataFramePreviewData[0]);
+        columnKeys.sort(sortColumns(groupByArray));
+        const tableColumns = columnKeys.map(k => {
+          return {
+            field: k,
+            name: k,
+            sortable: true,
+            truncateText: true,
+          };
+        });
+        const sorting = {
+          sort: {
+            field: tableColumns[0].field,
+            direction: SortDirection.ASC,
+          },
+        };
+        setIsLoading(false);
+        setColumns(tableColumns);
+        setSort(sorting);
+      }
+    },
+    [dataFramePreviewData]
```

Comment:
I'm afraid tracking the object here will not work as expected, `useEffect` isn't doing a deep equal (look [here](https://github.com/facebook/react/issues/14476#issuecomment-471199055) for other options). I wonder, could this whole `useEffect` block be just part of `getPreview`?

Reply:
Moved this into `getPreview` for simplification. So this second `useEffect` is not used anymore 
fd3c34164e5ead57a869c15608d253990f2a3f20